/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2018-2019. All rights reserved.
 * Description: dump file data
 * Author: dml
 * Create: 2018-06-13
 */
#include "rpc_dumper.h"

#include <algorithm>

#include "config/config.h"
#include "error_code.h"
#include "prof_params.h"
#include "msprof_dlog.h"
#include "rpc_data_handle.h"
#include "utils.h"
#include "param_validation.h"

namespace Msprof {
namespace Engine {
using namespace analysis::dvvp::common::error;
using namespace analysis::dvvp::common::config;
using namespace analysis::dvvp::common::utils;
using namespace analysis::dvvp::common::validation;
using namespace Analysis::Dvvp::MsprofErrMgr;
/**
* @brief RpcDumper: the construct function
* @param [in] module: the path of profiling data to be saved
*/
RpcDumper::RpcDumper(const std::string &module)
    : DataDumper(), moduleNameWithId_(module), hostPid_(-1), devId_(-1), dataHandle_(nullptr)
{
}

RpcDumper::~RpcDumper()
{
    Stop();
}

int RpcDumper::GetNameAndId(const std::string &module)
{
    MSPROF_LOGI("GetNameAndId module:%s", module.c_str());
    size_t posTmp = module.find_first_of("-");
    if (posTmp == std::string::npos) {
        module_ = module;
        return PROFILING_SUCCESS;
    }
    size_t pos = module.find_last_of("-");
    if (posTmp >= pos) {
        MSPROF_LOGE("get pos failed, module:%s", module.c_str());
        MSPROF_INNER_ERROR("EK9999", "get pos failed, module:%s", module.c_str());
        return PROFILING_FAILED;
    }
    module_ = module.substr(0, posTmp);
    std::string hostPidStr = module.substr(posTmp + 1, pos - posTmp - 1);
    try {
        hostPid_ = std::stoi(hostPidStr);
    } catch (...) {
        MSPROF_LOGE("Failed to transfer hostPidStr(%s) to integer.", hostPidStr.c_str());
        MSPROF_INNER_ERROR("EK9999", "Failed to transfer hostPidStr(%s) to integer.", hostPidStr.c_str());
        return PROFILING_FAILED;
    }
    std::string devIdStr = module.substr(pos + 1, module.size());
    try {
        devId_ = std::stoi(devIdStr);
    } catch (...) {
        MSPROF_LOGE("Failed to transfer hostPidStr(%s) to integer.", devIdStr.c_str());
        MSPROF_INNER_ERROR("EK9999", "Failed to transfer hostPidStr(%s) to integer.", devIdStr.c_str());
        return PROFILING_FAILED;
    }
    MSPROF_LOGI("Get module:%s, hostPid:%d, devId:%d success", module_.c_str(), hostPid_, devId_);
    return PROFILING_SUCCESS;
}

/**
* @brief Start: init variables of RpcDumper for can receive data from user plugin
*               start a new thread to check the data from user and write data to local files
* @return : success return PROFILING_SUCCESS, failed return PROFIING_FAILED
*/
int RpcDumper::Start()
{
    if (started_) {
        MSPROF_LOGW("this reporter has been started!");
        return PROFILING_SUCCESS;
    }
    int ret = GetNameAndId(moduleNameWithId_);
    if (ret != PROFILING_SUCCESS) {
        MSPROF_LOGE("RpcDumper GetNameAndId failed");
        MSPROF_INNER_ERROR("EK9999", "RpcDumper GetNameAndId failed");
        return PROFILING_FAILED;
    }
    ReceiveData::moduleName_ = module_;
    MSVP_MAKE_SHARED4_RET(dataHandle_, RpcDataHandle, moduleNameWithId_, module_, hostPid_, devId_, PROFILING_FAILED);
    ret = dataHandle_->Init();
    if (ret != PROFILING_SUCCESS) {
        MSPROF_LOGE("RpcDumper dataHandle init failed");
        MSPROF_INNER_ERROR("EK9999", "RpcDumper dataHandle init failed");
        return PROFILING_FAILED;
    }
    Thread::SetThreadName(analysis::dvvp::common::config::MSVP_RPC_DUMPER_THREAD_NAME);
    ret = Thread::Start();
    if (ret != PROFILING_SUCCESS) {
        MSPROF_LOGE("Failed to start the reporter %s in RpcDumper::Start().", moduleNameWithId_.c_str());
        MSPROF_INNER_ERROR("EK9999", "Failed to start the reporter %s in RpcDumper::Start().",
            moduleNameWithId_.c_str());
        return PROFILING_FAILED;
    } else {
        MSPROF_LOGI("Succeeded in starting the reporter %s in RpcDumper::Start().", moduleNameWithId_.c_str());
    }

    size_t capacity = RING_BUFF_CAPACITY;
    auto iter = std::find_if(std::begin(MSPROF_MODULE_ID_NAME_MAP), std::end(MSPROF_MODULE_ID_NAME_MAP),
        [&](ModuleIdName m) { return m.name == module_; });
    if (iter != std::end(MSPROF_MODULE_ID_NAME_MAP)) {
        capacity = iter->ringBufSize;
    }
    ret = ReceiveData::Init(capacity);
    if (ret != PROFILING_SUCCESS) {
        MSPROF_LOGE("ReceiveData Init failed");
        MSPROF_INNER_ERROR("EK9999", "ReceiveData Init failed");
        return PROFILING_FAILED;
    }
    started_ = true;
    MSPROF_LOGI("start reporter success. module:%s, capacity:%llu", module_.c_str(), capacity);
    return PROFILING_SUCCESS;
}

int RpcDumper::Report(CONST_REPORT_DATA_PTR rData)
{
    return DoReport(rData);
}

uint32_t RpcDumper::GetReportDataMaxLen()
{
    MSPROF_LOGI("GetReporterMaxLen from module: %s", module_.c_str());
    return RECEIVE_CHUNK_SIZE;
}

/**
* @brief Run: the thread function to deal with user datas
*/
void RpcDumper::Run(const struct error_message::Context &errorContext)
{
    MsprofErrorManager::instance()->SetErrorContext(errorContext);
    DoReportRun();
}

/**
* @brief Stop: wait data write finished, then stop the thread, which check data from user
*/
int RpcDumper::Stop()
{
    int ret = PROFILING_SUCCESS;
    if (started_) {
        started_ = false;
        ReceiveData::StopReceiveData();
        ret = Thread::Stop();
        if (ret != PROFILING_SUCCESS) {
            MSPROF_LOGE("Failed to stop the reporter %s in RpcDumper::Stop().", module_.c_str());
            MSPROF_INNER_ERROR("EK9999", "Failed to stop the reporter %s in RpcDumper::Stop().", module_.c_str());
            return PROFILING_FAILED;
        } else {
            MSPROF_LOGI("Succeeded in stopping the reporter %s in RpcDumper::Stop().", module_.c_str());
        }
        dataHandle_->UnInit();
        ReceiveData::PrintTotalSize();
        MSPROF_LOGI("RpcDumper stop module:%s", moduleNameWithId_.c_str());
    }
    return ret;
}

void RpcDumper::WriteDone()
{
    SHARED_PTR_ALIA<FileChunkReq> fileChunk;
    MSVP_MAKE_SHARED0_VOID(fileChunk, FileChunkReq);
    fileChunk->set_filename(module_);
    fileChunk->set_datamodule(analysis::dvvp::common::config::FileChunkDataModule::PROFILING_IS_FROM_MSPROF);
    fileChunk->set_islastchunk(true);
    fileChunk->set_needack(false);
    fileChunk->set_offset(-1);
    fileChunk->set_chunksizeinbytes(0);

    SHARED_PTR_ALIA<analysis::dvvp::message::JobContext> jobCtx = nullptr;
    MSVP_MAKE_SHARED0_VOID(jobCtx, analysis::dvvp::message::JobContext);
    jobCtx->module = module_;
    jobCtx->dataModule = analysis::dvvp::common::config::FileChunkDataModule::PROFILING_IS_FROM_MSPROF;
    MSPROF_LOGI("RpcDumper WriteDone for module %s", moduleNameWithId_.c_str());

    auto iter = devIds_.begin();
    for (; iter != devIds_.end(); iter++) {
        jobCtx->dev_id = *iter;
        fileChunk->mutable_hdr()->set_job_ctx(jobCtx->ToString());

        std::string data = analysis::dvvp::message::EncodeMessage(fileChunk);
        if (data.size() > 0) {
            if (dataHandle_->SendData(data.c_str(), data.size(), module_.c_str(), jobCtx->ToString().c_str())
                != PROFILING_SUCCESS) {
                MSPROF_LOGE("send last FileChunk package failed");
                MSPROF_INNER_ERROR("EK9999", "send last FileChunk package failed");
                continue;
            }
        } else {
            MSPROF_LOGE("Encode the last FileChunk failed");
            MSPROF_INNER_ERROR("EK9999", "Encode the last FileChunk failed");
            continue;
        }
    }
}

/**
* @brief Flush: write all datas from user to local files
* @return : success return PROFILING_SUCCESS, failed return PROFIING_FAILED
*/
int RpcDumper::Flush()
{
    if (!started_) {
        MSPROF_LOGW("this reporter has been stopped");
        return PROFILING_SUCCESS;
    }
    MSPROF_EVENT("[RpcDumper::Flush]Begin to flush data, module:%s", moduleNameWithId_.c_str());
    ReceiveData::Flush();
    TimedTask();
    MSPROF_LOGI("ReceiveData Flush finished");
    MSPROF_EVENT("[RpcDumper::Flush]End to flush data, module:%s", moduleNameWithId_.c_str());
    return PROFILING_SUCCESS;
}

void RpcDumper::TimedTask()
{
    dataHandle_->Flush();
}

/**
* @brief Dump: write the user datas in messages into local files
* @param [in] messages: the vector saved the user datas to be write to local files
* @return : success return PROFILING_SUCCESS, failed return PROFIING_FAILED
*/
int RpcDumper::Dump(std::vector<SHARED_PTR_ALIA<FileChunkReq>> &messages)
{
    int ret = PROFILING_SUCCESS;
    if (!(dataHandle_->IsReady()) && (messages.size() > 0)) {
        dataHandle_->TryToConnect();
    }
    for (size_t i = 0; i < messages.size(); i++) {
        if (messages[i] == nullptr) {
            continue;
        }
        std::string tag = messages[i]->tag();
        if (!ParamValidation::instance()->CheckDataTagIsValid(tag)) {
            MSPROF_LOGE("UploaderDumper::Dump, Check tag failed, module:%s, tag:%s",
                module_.c_str(), messages[i]->tag().c_str());
            MSPROF_INNER_ERROR("EK9999", "UploaderDumper::Dump, Check tag failed, module:%s, tag:%s",
                module_.c_str(), messages[i]->tag().c_str());
            continue;
        }
        messages[i]->set_datamodule(FileChunkDataModule::PROFILING_IS_FROM_MSPROF);
        // update total_size
        std::string encoded = analysis::dvvp::message::EncodeMessage(messages[i]);
        std::string fileName = module_ + "." + messages[i]->tag() + "." + messages[i]->tagsuffix();
        ret = dataHandle_->SendData(encoded.c_str(), encoded.size(), fileName, messages[i]->hdr().job_ctx().c_str());
        if (ret != PROFILING_SUCCESS) {
            MSPROF_LOGE("RpcDumper::Dump, UploadData failed, fileName:%s, chunkLen:%d",
                module_.c_str(), messages[i]->chunksizeinbytes());
            MSPROF_INNER_ERROR("EK9999", "RpcDumper::Dump, UploadData failed, fileName:%s, chunkLen:%d",
                module_.c_str(), messages[i]->chunksizeinbytes());
        }
    }
    return ret;
}
}
}
